src/domain/usecases/ValidateOptionSetsUseCase.ts

```diff
@@ -0,0 +1,106 @@
+import _ from "lodash";
+import { OptionSet } from "domain/entities/OptionSet";
+import { OptionValidationResult, OptionValidationRule } from "domain/entities/OptionValidationResult";
+import { OptionSetRepository } from "domain/repositories/OptionSetRepository";
+import logger from "utils/log";
+import { Async } from "domain/entities/Async";
+import { Option } from "domain/entities/Option";
+
+export class ValidateOptionSetsUseCase {
+    constructor(private readonly optionSetRepository: OptionSetRepository) {}
+
+    async execute(options: UseCaseOptions): Async<OptionValidationResult[]> {
+        logger.info("Loading optionSets...");
+        const optionsSets = await this.optionSetRepository.getAll();
+        return this.buildOptionValidationResults(optionsSets, options);
+    }
+
+    private buildOptionValidationResults(
+        optionsSets: OptionSet[],
+        options: UseCaseOptions
+    ): OptionValidationResult[] {
+        const duplicateOptions = this.getDuplicateOptions(optionsSets);
+        const validationResults = this.buildValidationResults(optionsSets, options);
+        return duplicateOptions.concat(validationResults).filter(vr => vr.rules.length > 0);
+    }
+
+    private buildValidationResults(
+        optionSets: OptionSet[],
+        options: UseCaseOptions
+    ): OptionValidationResult[] {
+        const allOptions = optionSets.flatMap(optionSet =>
+            optionSet.options.map((option, optionIndex) => ({ option, optionSet, optionIndex }))
+        );
+
+        return allOptions.map(({ option, optionIndex, optionSet }): OptionValidationResult => {
+            const invalidLengthRules = this.validateLength(option, options.codeLength);
+            const namingConventionRules = this.validateNamingConventions(option);
+            const misplacedCommasRules = this.validateMisplacedCommas(option);
+            const orderRules = this.validateOrder(option, optionIndex);
+            return {
+                option,
+                optionSet,
+                rules: [
+                    ...invalidLengthRules,
+                    ...namingConventionRules,
+                    ...misplacedCommasRules,
+                    ...orderRules,
+                ],
+            };
+        });
+    }
+
+    private validateLength(option: Option, validationLength: number): OptionValidationRule[] {
+        return option.code.length > validationLength
+            ? [{ type: "invalid_length", message: `Code length is ${option.code.length}` }]
+            : [];
```

Comment:
This check conceptually will never be returning N>1 validations, right? then the signature should reflect it: `Maybe<OptionValidationRule>`